R/add_hri_variables.R

```diff
@@ -0,0 +1,137 @@
+
+  # Create a flag, 'keep_flag', to determine whether individuals are Scottish
+  # residents or not
+  return_data <- data %>%
+    dplyr::mutate(pc_area = stringr::str_match(postcode, "^[A-Z]{1,3}")) %>%
+    dplyr::left_join(pc_areas, by = "pc_area") %>%
+    dplyr::mutate(
+      dummy_postcode = .data$postcode %in% c("BF010AA", "NF1 1AB", "NK010AA") |
+        stringr::str_sub(.data$postcode, 1, 4) %in% c("ZZ01", "ZZ61"),
+      eng_prac = .data$gpprac %in% c(99942, 99957, 99961, 99976, 99981, 99995, 99999),
+      keep_flag = dplyr::case_when(
+        .data$scot_flag == 0L ~ 0L,
+        (is_missing(.data$postcode) | .data$dummy_postcode) & is.na(.data$gpprac) ~ 2L,
+        !.data$eng_prac & is_missing(.data$postcode) ~ 3L,
+        !.data$eng_prac & .data$dummy_postcode ~ 4L,
+        TRUE ~ 1L
+      )
+    ) %>%
+    dplyr::select(-"dummy_postcode", -"eng_prac")
+
+  return(return_data)
+}
+
+#' Add HRI variables to an SLF Individual File
+#'
+#' @details Filters the dataset to only include Scottish residents, then
+#' creates a lookup where HRIs are calculated at Scotland, Health Board, and
+#' LCA level. Then joins on this lookup by chi/anon_chi.
+#'
+#' @param data An SLF individual file
+#' @param chi_variable Defaults to "anon_chi"
+#'
+#' @return The individual file with HRI variables matched on
+#' @export
+add_hri_variables <- function(data,
+                              chi_variable = "anon_chi") {
+  hri_lookup <- data %>%
+    dplyr::select(
+      "year",
+      chi_variable,
+      "postcode",
+      "gpprac",
+      "lca",
+      "hbrescode",
+      "health_net_cost",
+      "acute_episodes",
+      "mat_episodes",
+      "mh_episodes",
+      "gls_episodes",
+      "op_newcons_attendances",
+      # op_newcons_dnas,
+      "ae_attendances",
+      "pis_paid_items",
+      "ooh_cases"
+    ) %>%
+    flag_non_scottish_residents() %>%
+    dplyr::filter(keep_flag == 0L) %>%
+    # Scotland cost and proportion
+    dplyr::mutate(
+      scotland_cost = sum(health_net_cost),
+      scotland_pct = (health_net_cost / scotland_cost) * 100
+    ) %>%
+    dplyr::arrange(dplyr::desc(health_net_cost)) %>%
+    dplyr::mutate(hri_scotp = cumsum(scotland_pct)) %>%
+    # Health Board
+    dplyr::group_by(hbrescode) %>%
+    dplyr::mutate(
+      hb_cost = sum(health_net_cost),
+      hb_pct = (health_net_cost / hb_cost) * 100
+    ) %>%
+    dplyr::arrange(dplyr::desc(health_net_cost), .by_group = TRUE) %>%
+    dplyr::mutate(hri_hbp = cumsum(hb_pct)) %>%
+    dplyr::ungroup() %>%
+    # LCA
+    dplyr::group_by(lca) %>%
+    dplyr::mutate(
+      lca_cost = sum(health_net_cost),
+      lca_pct = (health_net_cost / lca_cost) * 100
+    ) %>%
+    dplyr::arrange(dplyr::desc(health_net_cost), .by_group = TRUE) %>%
+    dplyr::mutate(hri_lcap = cumsum(lca_pct)) %>%
+    dplyr::ungroup() %>%
+    # Add HRI flags
+    dplyr::mutate(
+      hri_scot = hri_scotp <= 50L,
+      hri_hb = hri_hbp <= 50L,
+      hri_lca = hri_lcap <= 50L,
```

Comment:
Good work for being explicit, but these aren't actually integers, so we should be specific like this:
```suggestion
      hri_scot = hri_scotp <= 50.0,
      hri_hb = hri_hbp <= 50.0,
      hri_lca = hri_lcap <= 50.0,
```